### PR TITLE
Allow End of month/year for future date selections

### PR DESCRIPTION
**Pull request by liedekef**
This change allows a new selection "End of month" and "End of year" (like "Beginning of month" and "Beginning of year" already exist).

| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #number

### CODE REVIEW

File: src/Html.php [hunk 1]
```diff
@@ -3408,10 +3408,18 @@ public static function getGenericDateTimeSearchItems($options)
                 $dates[$i . 'WEEK'] = sprintf(_n('+ %d week', '+ %d weeks', $i), $i);
             }
 
+            if ($params['with_days']) {
+                    $dates['ENDMONTH']  = __('End of the month');
```
Comment (cedric-anne) [inline src/Html.php:3412]:
```suggestion
                $dates['ENDMONTH']  = __('End of the month');
```

File: src/Html.php [hunk 2]
```diff
@@ -3408,10 +3408,18 @@ public static function getGenericDateTimeSearchItems($options)
...
+            }
+
             for ($i = 1; $i <= 12; $i++) {
                 $dates[$i . 'MONTH'] = sprintf(_n('+ %d month', '+ %d months', $i), $i);
             }
 
+            if ($params['with_days']) {
+                    $dates['ENDYEAR']  = __('End of the year');
```
Comment (cedric-anne) [inline src/Html.php:3420]:
```suggestion
                $dates['ENDYEAR']  = __('End of the year');
```

File: src/Html.php [hunk 3]
```diff
@@ -3474,6 +3482,29 @@ public static function computeGenericDateTimeSearch($val, $force_day = false, $s
...
+                    $day = 31;
+                    break;
+
+                case "ENDMONTH":
+                    $day = date("t", $specifictime);
+                    break;
+            }
+
+            return date($format_use, mktime($hour, $minute, $second, $month, $day, $year));
+        }
```
Comment (cedric-anne) [inline src/Html.php:3506]: This should be mutualized with the `if (strstr($val, 'BEGIN')) {`.

```php
if (strstr($val, 'BEGIN') || strstr($val, 'END')) {
    ...
    switch ($val) {
        ...
        case "ENDYEAR":
            $month = 12;
            $day = 31;
            break;

        case "ENDMONTH":
            $day = date("t", $specifictime);
            break;
```

**Review by cedric-anne — CHANGES_REQUESTED** [review]

File: src/Html.php (hunk 1 above)
Comment (liedekef) [reply src/Html.php]: Done in new code

File: src/Html.php (hunk 2 above)
Comment (liedekef) [reply src/Html.php]: Done in new code

File: src/Html.php (hunk 3 above)
Comment (liedekef) [reply src/Html.php]: Done in new code

**Comment by liedekef** [comment]
Code updated, please re-review

**Review by cedric-anne — APPROVED** [review]

**Review by orthagh — APPROVED** [review]

**Review by trasher — APPROVED** [review]

**Pull request merged**